spec/formatting.md

```diff
@@ -1,24 +1,124 @@
-# WIP DRAFT MessageFormat 2.0 Formatting Behaviour
+# DRAFT MessageFormat 2.0 Formatting
 
 ## Introduction
 
 This document defines the behaviour of a MessageFormat 2.0 implementation
 when formatting a message for display in a user interface, or for some later processing.
 
-The document is part of the MessageFormat 2.0 specification,
-the successor to ICU MessageFormat, henceforth called ICU MessageFormat 1.0.
+To start, we presume that a _message_ has either been parsed from its syntax
+or created from a data model description.
+If this construction has encountered any Syntax or Data Model errors,
+their handling during formatting is specified here as well.
 
-## Literal Resolution
+Formatting of a _message_ is defined by the following operations:
 
-The resolved value of a _text_ or a _literal_ is the character sequence of the _text_
-or _literal_ after any character escape has been converted to the escaped character.
-When a _literal_ is used as an _expression_ argument or on the right-hand side of an _option_,
-the formatting function MUST treat their resolved values the same independently of whether the
-value was originally _quoted_ or _unquoted_.
-For example, the _option_ `foo=42` and the _option_ `foo=|42|` 
-are treated as identical.
+- **_Resolution_** determines the value of a part of the message,
+  such as an _expression_, with reference to the current _formatting context_.
+  This can include multiple steps,
+  such as looking up the value of a variable and calling formatting functions.
+  The resolved value is not necessarily in the shape it will finally take,
+  but is "formattable", i.e. it contains everything required by the eventual formatting.
 
-## Variable Resolution
+  > For example, the resolved value of the _expression_ `{|0.40| :number style=percent}`
+  > could be an object such as
+  >
+  > ```
+  > { value: Number('0.40'),
+  >   formatter: NumberFormat(locale, { style: 'percent' }) }
+  > ```
+  >
+  > The shape of this value is implementation-dependent,
+  > and different implementations may choose to perform different levels of resolution.
+
+- **_Pattern Selection_** determines which of a message's _patterns_ is formatted.
+  For a message with no _selectors_, this is simple as there is only one _pattern_.
+  With _selectors_, this will depend on their _resolution_.
+
+- **_Formatting_** takes the resolved values of the selected _pattern_,
```

Comment:
The earlier definition of "resolution" said "**_Resolution_** determines the value of a part of the message", but this definition of "formatting" implies that the output of resolution is one or more values. So for consistency, either the earlier definition should say "values" instead of "value", or this definition should rely on a shared definition of "value" that includes multiple parts.

Reply:
The plural here is intended to refer to the distinct _text_ and _expression_ parts of a _pattern_. Each such part of _pattern_ would still resolve to one value each.

Reply:
In that case, maybe it's worth adding something after line 15 saying that the result of resolving a pattern is a list of values that results from independently resolving each of its parts? If the definition of formatting a pattern requires a list as input, then maybe it's worth saying that the output of resolving a pattern is a list, even if the shape of a "value" is being left abstract.

Reply:
In that case, maybe it's worth adding something after line 15 saying that the result of resolving a pattern is a list of values that results from independently resolving each of its parts? If the definition of formatting a pattern requires a list as input, then maybe it's worth saying that the output of resolving a pattern is a list, even if the shape of a "value" is being left abstract.

Reply:
Your comments lead me to realise that really practically all of the "resolution" section is about "expression resolution". So I'm retitling accordingly, hopefully adding some clarity here as well.